src/session_pool/session_pool.js

```diff
@@ -0,0 +1,314 @@
+        return this.sessions.filter(session => session.isUsable()).length;
+    }
+
+    /**
+     * Gets count of blocked sessions in the pool.
+     * @return {number}
+     */
+    get retiredSessionsCount() {
+        return this.sessions.filter(session => !session.isUsable()).length;
+    }
+
+    /**
+     * Starts periodic state persistence and potentially loads SessionPool state from {@link KeyValueStore}.
+     * This function must be called before you can start using the instance in a meaningful way.
+     *
+     * @return {Promise<void>}
+     */
+    async initialize() {
+        this.keyValueStore = await openKeyValueStore(this.persistStateKeyValueStoreId);
+
+        // in case of migration happened and SessionPool state should be restored from the keyValueStore.
+        await this._maybeLoadSessionPool();
+
+        events.on(ACTOR_EVENT_NAMES_EX.PERSIST_STATE, this.persistState.bind(this));
+    }
+
+    /**
+     * Gets session.
+     * If there is space for new session, it creates and return new session.
+     * If the session pool is full, it picks a session from the pool,
+     * If the picked session is usable it is returned, otherwise it creates and returns a new one.
+     *
+     * @return {Promise<Session>}
+     */
+    async retrieveSession() {
+        // If we have enough space for session. Return newly created session.
+        if (this._hasSpaceForSession()) {
+            return this._createSession();
+        }
+
+        // For example that developer can plug different picking algorithms such as the Lukášův
+        // Maybe a should have pickSession function to be customizable.
+        const pickedSession = this._pickSession();
+
+        // If session can be used return the session
+        if (pickedSession.isUsable()) {
+            return pickedSession;
+        }
+
+        //  otherwise remove old session and return newly created session
+        this._removeSession(pickedSession);
+        return this._createSession();
+    }
+
+    /**
+     * Returns an object representing the internal state of the `SessionPool` instance.
+     * Note that the object's fields can change in future releases.
+     *
+     * @returns {Object}
+     */
+    getState() {
+        return {
+            usableSessionsCount: this.usableSessionsCount,
+            retiredSessionsCount: this.retiredSessionsCount,
+            sessions: this.sessions.map(session => session.getState()),
+        };
+    }
+
+    /**
+     * Persists the current state of the `SessionPool` into the default {@link KeyValueStore}.
+     * The state is persisted automatically in regular intervals.
+     *
+     * @return {Promise}
+     */
+    async persistState() {
+        log.debug('SessionPool: Persisting state');
+        log.debug(`SessionPool: persistStateKeyValueStoreId: ${this.persistStateKeyValueStoreId}, persistStateKey: ${this.persistStateKey} `);
+        return this.keyValueStore.setValue(this.persistStateKey, this.getState());
```

Comment:
I'd `await this.keyValueStore ...` instead of returning. We're not interested in the return value of the call, whatever it may be.